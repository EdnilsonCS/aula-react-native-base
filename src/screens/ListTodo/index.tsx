import { useTodo } from '@hooks/todo';
import { useNavigation } from '@react-navigation/native';
import React, { useCallback } from 'react';
import { View } from 'react-native';
import {
  TodoList,
  Container,
  Title,
  TextDescription,
  Header,
  Icon,
  TitleItem,
  PlusButton,
} from './styles';

export interface ItemTodo {
  id: string;
  title: string;
  description: string;
}

const ListTodo: React.FC = () => {
  const { todo } = useTodo();
  const navigation = useNavigation();
  const renderItem = ({ item }: { item: ItemTodo }): React.ReactElement => {
    return (
      <View
        key={item.id}
        testID={item.id}
        style={{ borderColor: 'black', borderWidth: 1, padding: 20 }}
      >
        <TitleItem>Titulo: {item.title}</TitleItem>
        <TextDescription>Descrição: {item.description}</TextDescription>
      </View>
    );
  };
  const handleToNavigationToNewItem = useCallback(() => {
    navigation.navigate('CreateItemTodo');
  }, [navigation]);
  return (
    <Container>
      <Header>
        <Title>Lista de itens</Title>
        <PlusButton
          onPress={handleToNavigationToNewItem}
          testID="new-create-item"
        >
          <Icon size={40} name="plus-circle" />
        </PlusButton>
      </Header>
      <TodoList renderItem={renderItem} data={todo} />
    </Container>
  );
};

export default ListTodo;
