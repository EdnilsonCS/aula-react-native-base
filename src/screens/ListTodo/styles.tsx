import styled from 'styled-components/native';
import { FlatList } from 'react-native';
import FeatherIcon from 'react-native-vector-icons/Feather';
import { colors } from '@styles/index';
import { getStatusBarHeight } from 'react-native-iphone-x-helper';
import { ItemTodo } from './index';

export const Title = styled.Text`
  font-size: 20px;
`;

export const Container = styled.View`
  padding-top: ${getStatusBarHeight()}px;
`;

export const TodoList = styled(FlatList as new () => FlatList<ItemTodo>)`
  padding: 23px 24px 16px;
`;

export const TitleItem = styled.Text``;

export const TextDescription = styled.Text``;

export const Header = styled.View`
  flex-direction: row;
  align-items: center;
  justify-content: center;
  position: relative;
  height: 90px;
  background-color: ${colors.orange};
`;

export const Icon = styled(FeatherIcon)``;

export const PlusButton = styled.TouchableOpacity`
  position: absolute;
  right: 10px;
`;
