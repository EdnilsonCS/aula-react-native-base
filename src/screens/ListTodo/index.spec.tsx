import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import ListTodo from './index';

const mockedHistoryPush = jest.fn();

jest.mock('@react-navigation/native', () => {
  return {
    useNavigation: () => ({
      goBack: jest.fn(),
      navigate: mockedHistoryPush,
    }),
  };
});

const mockTodo = [
  {
    id: 'teste-id-1',
    title: 'title-fake-1',
    description: 'description-fake-2',
  },
  {
    id: 'teste-id-2',
    title: 'title-fake-1',
    description: 'description-fake-2',
  },
];

jest.mock('../../hooks/todo.tsx', () => {
  return {
    useTodo: () => ({
      todo: mockTodo,
    }),
  };
});

describe('List Todo', () => {
  beforeEach(() => {
    mockedHistoryPush.mockClear();
  });
  it('Should be able to render correct', () => {
    const tree = render(<ListTodo />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should be able to handle list of todo ', () => {
    const { getByTestId } = render(<ListTodo />);

    expect(getByTestId('teste-id-1')).toBeTruthy();
    expect(getByTestId('teste-id-2')).toBeTruthy();
  });

  it('Should ne able to navigate to createItemTodo', () => {
    const { getByTestId } = render(<ListTodo />);
    const buttonNewField = getByTestId('new-create-item');
    fireEvent.press(buttonNewField);
    expect(mockedHistoryPush).toHaveBeenCalledWith('CreateItemTodo');
  });
});
