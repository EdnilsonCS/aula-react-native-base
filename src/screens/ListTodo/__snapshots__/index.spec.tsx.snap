// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Todo Should be able to render correct 1`] = `
<View
  style={
    Array [
      Object {
        "paddingTop": 20,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-end",
          "backgroundColor": "#FFA800",
          "height": 90,
          "justifyContent": "center",
          "paddingRight": 20,
        },
      ]
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="new-create-item"
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 40,
            },
            Array [
              Object {},
            ],
            Object {
              "fontFamily": "Feather",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <RCTScrollView
    data={
      Array [
        Object {
          "description": "description-fake-2",
          "id": "teste-id-1",
          "title": "title-fake-1",
        },
        Object {
          "description": "description-fake-2",
          "id": "teste-id-2",
          "title": "title-fake-1",
        },
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "paddingBottom": 16,
          "paddingLeft": 24,
          "paddingRight": 24,
          "paddingTop": 23,
        },
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderColor": "black",
              "borderWidth": 1,
              "padding": 20,
            }
          }
          testID="teste-id-1"
        >
          <Text
            style={
              Array [
                Object {},
              ]
            }
          >
            Titulo: 
            title-fake-1
          </Text>
          <Text
            style={
              Array [
                Object {},
              ]
            }
          >
            Descrição: 
            description-fake-2
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "borderColor": "black",
              "borderWidth": 1,
              "padding": 20,
            }
          }
          testID="teste-id-2"
        >
          <Text
            style={
              Array [
                Object {},
              ]
            }
          >
            Titulo: 
            title-fake-1
          </Text>
          <Text
            style={
              Array [
                Object {},
              ]
            }
          >
            Descrição: 
            description-fake-2
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
